modules/internal/DefaultRectangular.chpl

```diff
@@ -2251,13 +2251,14 @@ module DefaultRectangular {
     use RangeChunk;
 
     type resType = op.generate().type;
-    var res: [dom] resType;
+    var res = dom.buildArray(resType, initElts=!isPOD(resType));
```

Comment:
Two questions:

1) Was the user-facing `noinit` feature not applicable here, or did it not give similar results?  (i.e., why call `dom.buildArray()` rather than using `noinit`?)

2) What do you think about rewriting `this.chpl__preScan()` to return `res` so that we could rely on split initialization rather than `noinit` + late registration by writing, say:

```chapel
res = this.chpl__preScan(op, res, dom, ...);
```

where `...` would be the other things that chpl__preScan currently returns (which might be awkward, admittedly...).  Or, I suppose nothing about that would need to be split initialization...  we could just put it in the initializer expression slot for the variable declaration.

Even better would be if could we say:

```chapel
(res, numTasks, rngs, state, _) = this.chpl_preScan(op, res, dom);
```

and have that get treated as its initialization, but this doesn't seem to work today.  :(  Tagging @mppf to see whether he thinks, with a bit more effort, we could have tuple-based assignment get treated as split initialization.

[Since this isn't user facing code, I don't consider either of these questions to be worth holding up a performance optimization for; but I think they're worth asking because if we want to do it and would benefit, then a user would as well.]

Reply:
See the last commit and https://github.com/chapel-lang/chapel/pull/16919#discussion_r554057807 for why this uses `buildArray` instead of `noinit`. TL;DR is that noinit only works for POD types. e.g.:

```chpl
record nonPod { proc deinit() { } }
var A: [1..10] nonPod = noinit;
```

```
nonPodNoinit.chpl:2: error: noinit is only supported for arrays of trivially copyable types
```

For chpl__preScan, I initially liked this idea, but when I went looking a little more I think this would just push the noinit down into `chpl__preScan`, but we'd still need a noinit somewhere.

Reply:
> Tagging @mppf to see whether he thinks, with a bit more effort, we could have tuple-based assignment get treated as split initialization.

This seems like something we can get working and deserves an issue of its own.

Reply:
> but when I went looking a little more I think this would just push the noinit down into chpl__preScan, but we'd still need a noinit somewhere.

Ah, good point.  Bummer.

Reply:
Thanks @mppf.  I added issue #16937.